sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -486,9 +486,17 @@ object TypeCoercion {
         }
 
       case i @ In(a, b) if b.exists(_.dataType != a.dataType) =>
-        findWiderCommonType(i.children.map(_.dataType)) match {
-          case Some(finalDataType) => i.withNewChildren(i.children.map(Cast(_, finalDataType)))
-          case None => i
+        if (b.map(_.dataType).distinct.size == 1) {
```

Comment:
mmh...this means that: `1 in ('1', 1)` behaves differently from `'1' in (1, 1)` which is not great IMHO. Can you please also check which is the behavior with nested subqueries? I think that having the same behavior among IN with literals and IN with subqueries is even more important that having the same behavior with binary comparisons.